### DISCUSSION TITLE
Aliased generic union with ParamSpec using stringified Ellipsis erros out

### DISCUSSION THREAD

**Discussion by Bluenix2** [python/typing#1169] [Q&A]
## Summary

I have two code examples using mixes of ParamSpec, Unions, and Ellipsis causing runtime errors.

### Stringified ellipsis in generic union

```python
from typing import Generic, ParamSpec, Union, TypeVar

P = ParamSpec('P')
RT = TypeVar('RT')

class TypeA(Generic[P, RT]): ...
class TypeB(Generic[P, RT]): ...

Alias = Union[TypeA[P, RT], TypeB[P, RT]]
Alias['...', object]
```
```
Traceback (most recent call last):
  File "<string>", line 10, in <module>
  File "/usr/local/lib/python3.10/typing.py", line 311, in inner
    return func(*args, **kwds)
  File "/usr/local/lib/python3.10/typing.py", line 1061, in __getitem__
    arg = arg[subargs]
  File "/usr/local/lib/python3.10/typing.py", line 311, in inner
    return func(*args, **kwds)
  File "/usr/local/lib/python3.10/typing.py", line 1053, in __getitem__
    raise TypeError(f"Expected a list of types, an ellipsis, "
TypeError: Expected a list of types, an ellipsis, ParamSpec, or Concatenate. Got ForwardRef('...')
```

This could be fixed by stringifying the whole thing (`'Alias[..., object]'`).

### Getting typehints of stringified ellipsis in generic

In the same vein as the above (but stemming from a difference place), if you try to get the type hints of a normal generic with a stringified ellipsis in the place of the ParamSpec you get an error:

```python
from typing import Generic, ParamSpec, Union, TypeVar, get_type_hints

P = ParamSpec('P')
RT = TypeVar('RT')

class TypeA(Generic[P, RT]): ...

def example(param: TypeA['...', object]) -> None: ...
get_type_hints(example)
```
```
Traceback (most recent call last):
  File "<string>", line 9, in <module>
  File "/usr/local/lib/python3.10/typing.py", line 1849, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/usr/local/lib/python3.10/typing.py", line 328, in _eval_type
    ev_args = tuple(_eval_type(a, globalns, localns, recursive_guard) for a in t.__args__)
  File "/usr/local/lib/python3.10/typing.py", line 328, in <genexpr>
    ev_args = tuple(_eval_type(a, globalns, localns, recursive_guard) for a in t.__args__)
  File "/usr/local/lib/python3.10/typing.py", line 326, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/usr/local/lib/python3.10/typing.py", line 690, in _evaluate
    type_ = _type_check(
  File "/usr/local/lib/python3.10/typing.py", line 175, in _type_check
    raise TypeError(f"{msg} Got {arg!r:.100}.")
TypeError: Forward references must evaluate to types. Got Ellipsis.
```

---

My only question is whether I am doing something wrong? The reason for this stringifying is compatibility but I am starting to run out of options with these issues.

**Comment by JelleZijlstra** [comment]
I don't think PEP 612 technically allows using ellipsis in this context. 3.11 will allow some similar things at runtime (https://github.com/python/cpython/commit/81120b6754c67a8f1f00cbc3af6963c0e1945911), but it's still not a well-specified area.

**Reply by erictraut** [reply]
I think PEP 612 does allow ellipsis as a type argument for a ParamSpec within a generic type alias. At least that's how I interpreted the PEP when I implemented it in pyright. I think pyre also accepts this. It looks like mypy doesn't currently support it though. In fact, it looks like mypy doesn't like the use of a ParamSpec at all within a type alias (regardless of whether it uses an explicit `TypeAlias` annotation).

I'm not sure whether a quoted ellipsis is meant to be allowed. Python 3.10 seems to be fine with it in a `Callable` but not within the specialization of a generic type alias that refers to a `Union`. Pyright is happy with a quoted ellipsis in either of these places, and mypy generates an error in both cases. Definitely inconsistent. And for that reason, it's probably best to avoid the use of quoted ellipsis.

@Bluenix2, you said that you're using a quoted ellipsis for compatibility reasons. Can you elaborate? Are you talking about compatibility with older stdlib implementations? With `typing_extensions`? With mypy or other type checkers?

**Reply by Bluenix2** [reply]
@JelleZijlstra how else would I pass some value to mean "any parameters" in this union?

---

Python version compatibility, yes @erictraut, on non-Python 3.10 I got errors doing `Alias[..., object]` so I started by quoting just the ellipsis (`Alias['...', object]`) but then ran into the first issue I mentioned in my post on all versions (both Python 3.10 and at least Python 3.8 which is what I believe I use locally) so I ended up quoting everything.
